File: geometry/golang/vectors/vectors.go
package vectors

import (
	"didact/geometry/golang/utils"
	"fmt"
	"math"
)

// A two-dimensional cartesian vector. Represented by a two-part array.
type Vec2 [2]float64

// A three-dimensional cartesian vector. Represented by a three-part array.
type Vec3 [3]float64

// As opposed to operator "==", function "ApproxEq" tests whether
// floating-point, 2D vector components are approximately equal.
// Check "didact/geometry/golang/utils/approx-eq.go" for details.
func (v1 Vec2) ApproxEq(v2 Vec2) bool {
	eq := utils.ApproxEq
	for i := range v1 {
		if !eq(v1[i], v2[i]) {
			return false
		}
	}
	return true
}

// As opposed to operator "==", function "ApproxEq" tests whether
// floating-point, 3D vector components are approximately equal.
func (v1 Vec3) ApproxEq(v2 Vec3) bool {
	eq := utils.ApproxEq
	for i := range v1 {
		if !eq(v1[i], v2[i]) {
			return false
		}
	}
	return true
}

// Returns a 2D vector with the absolute values of its components.
func (v Vec2) Abs() Vec2 {
	for i := range v {
		v[i] = math.Abs(v[i])
	}
	return v
}

// Returns a 3D vector with the absolute values of its components.
func (v Vec3) Abs() Vec3 {
	for i := range v {
		v[i] = math.Abs(v[i])
	}
	return v
}

// Returns a 2D vector that is the sum of two 2D vectors.
func (v1 Vec2) Add(v2 Vec2) Vec2 {
	for i := range v1 {
		v1[i] += v2[i]
	}
	return v1
}

// Returns a 3D vector that is the sum of two 3D vectors.
func (v1 Vec3) Add(v2 Vec3) Vec3 {
	for i := range v1 {
		v1[i] += v2[i]
	}
	return v1
}

// Returns a 2D vector that is the difference of two 2D vectors.
func (v1 Vec2) Sub(v2 Vec2) Vec2 {
	for i := range v1 {
		v1[i] -= v2[i]
	}
	return v1
}

// Returns a 3D vector that is the difference of two 3D vectors.
func (v1 Vec3) Sub(v2 Vec3) Vec3 {
	for i := range v1 {
		v1[i] -= v2[i]
	}
	return v1
}

// Flips the signs of the 2D vector components.
func (v Vec2) Negate() Vec2 {
	for i := range v {
		v[i] = -v[i]
	}
	return v
}

// Flips the signs of the 3D vector components.
func (v Vec3) Negate() Vec3 {
	for i := range v {
		v[i] = -v[i]
	}
	return v
}

// Inverts the 2D vector components.
func (v Vec2) Invert() Vec2 {
	for i := range v {
		v[i] = 1.0 / v[i]
	}
	return v
}

// Inverts the 3D vector components.
func (v Vec3) Invert() Vec3 {
	for i := range v {
		v[i] = 1.0 / v[i]
	}
	return v
}

// Returns the magnitude of a 2D vector.
func (v Vec2) Mag() float64 {
	accum := 0.0
	for i := range v {
		accum += math.Pow(v[i], 2)
	}
	return math.Sqrt(accum)
}

// Returns the magnitude of a 3D vector.
func (v Vec3) Mag() float64 {
	accum := 0.0
	for i := range v {
		accum += math.Pow(v[i], 2)
	}
	return math.Sqrt(accum)
}

// Multiplies a 2D vector by a scalar.
func (v Vec2) Scale(s float64) Vec2 {
	for i := range v {
		v[i] *= s
	}
	return v
}

// Multiplies a 3D vector by a scalar.
func (v Vec3) Scale(s float64) Vec3 {
	for i := range v {
		v[i] *= s
	}
	return v
}

// Computes the dot product of two 2D vectors.
func (v1 Vec2) Dot(v2 Vec2) float64 {
	accum := 0.0
	for i := range v1 {
		accum += v1[i] * v2[i]
	}
	return accum
}

// Computes the dot product of two 3D vectors.
func (v1 Vec3) Dot(v2 Vec3) float64 {
	accum := 0.0
	for i := range v1 {
		accum += v1[i] * v2[i]
	}
	return accum
}

// Calculates the distance between two 2D vector points.
func (v1 Vec2) Distance(v2 Vec2) float64 {
	accum := 0.0
	for i := range v1 {
		accum += math.Pow(v2[i]-v1[i], 2)
	}
	return math.Sqrt(accum)
}

// Calculates the distance between two 3D vector points.
func (v1 Vec3) Distance(v2 Vec3) float64 {
	accum := 0.0
	for i := range v1 {
		accum += math.Pow(v2[i]-v1[i], 2)
	}
	return math.Sqrt(accum)
}

// Interpolates the components of the two 2D vectors.
func (v1 Vec2) Lerp(t float64, v2 Vec2) Vec2 {
	for i := range v1 {
		v1[i] += (v2[i] - v1[i]) * t
	}
	return v1
}

// Interpolates the components of the two 3D vectors.
func (v1 Vec3) Lerp(t float64, v2 Vec3) Vec3 {
	for i := range v1 {
		v1[i] += (v2[i] - v1[i]) * t
	}
	return v1
}

// Returns the unit vector of a 2D vector.
func (v Vec2) Normalize() Vec2 {
	mag := v.Mag()
	for i := range v {
		v[i] /= mag
	}
	return v
}

// Returns the unit vector of a 3D vector.
func (v Vec3) Normalize() Vec3 {
	mag := v.Mag()
	for i := range v {
		v[i] /= mag
	}
	return v
}

// Returns a 2D vector with its components rounded.
func (v Vec2) Round() Vec2 {
	for i := range v {
		v[i] = math.Round(v[i])
	}
	return v
}

// Returns a 3D vector with its components rounded.
func (v Vec3) Round() Vec3 {
	for i := range v {
		v[i] = math.Round(v[i])
	}
	return v
}

// Fulfills the Stringer interface for Vec2.
func (v Vec2) String() string {
	x := v[0]
	y := v[1]
	return fmt.Sprintf("vec2(%.2f, %.2f)", x, y)
}

// Fulfills the Stringer interface for Vec3.
func (v Vec3) String() string {
	x := v[0]
	y := v[1]
	z := v[2]
	return fmt.Sprintf("vec3(%.2f, %.2f, %.2f)", x, y, z)
}

File: geometry/julia/Vectors/src/Vectors.jl
module Vectors

import Base: +, -, *

export Vec2, Vec3, add, +, sub, -, neg, invert, mag, scale, *, dot, distance, normalize

"""
A cartesian representation of a vector in two dimensions.
"""
struct Vec2
    x::Float64
    y::Float64
end

"""
A cartesian representation of a vector in three dimensions.
"""
struct Vec3
    x::Float64
    y::Float64
    z::Float64
end

"""Computes the sum of two vectors."""
add(v1::Vec2, v2::Vec2) = Vec2(v1.x + v2.x, v1.y + v2.y)
add(v1::Vec3, v2::Vec3) = Vec3(v1.x + v2.x, v1.y + v2.y, v1.z + v2.z)
(+)(v1, v2) = add(v1, v2)
(+)(v) = v

"""Computes the difference of two vectors."""
sub(v1::Vec2, v2::Vec2) = Vec2(v1.x - v2.x, v1.y - v2.y)
sub(v1::Vec3, v2::Vec3) = Vec3(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z)
(-)(v1, v2) = sub(v1, v2)

"""Flips the signs of the vector components."""
neg(v::Vec2) = Vec2(-v.x, -v.y)
neg(v::Vec3) = Vec3(-v.x, -v.y, -v.z)
(-)(v) = neg(v)

"""Inverts the vector components."""
invert(v::Vec2) = Vec2(1 / v.x, 1 / v.y)
invert(v::Vec3) = Vec3(1 / v.x, 1 / v.y, 1 / v.z)

"""Computes the magnitude (a.k.a. length) of a vector."""
mag(v::Vec2) = hypot(v.x, v.y)
mag(v::Vec3) = hypot(v.x, v.y, v.z)

"""Scales a vector by a number."""
scale(v::Vec2, scalar::Real) = Vec2(v.x * scalar, v.y * scalar)
scale(v::Vec3, scalar::Real) = Vec3(v.x * scalar, v.y * scalar, v.z * scalar)
(*)(v, scalar) = scale(v, scalar)
(*)(scalar, v) = scale(v, scalar)

"""Computes the dot product of two vectors."""
dot(v1::Vec2, v2::Vec2) = (v1.x * v2.x) + (v1.y * v2.y)
dot(v1::Vec3, v2::Vec3) = (v1.x * v2.x) + (v1.y * v2.y) + (v1.z * v2.z)

"""Computes the distance between the tips of two vectors."""
distance(v1::Vec2, v2::Vec2) = hypot(v2.x - v1.x, v2.y - v1.y)
distance(v1::Vec3, v2::Vec3) = hypot(v2.x - v1.x, v2.y - v1.y, v2.z - v1.z)

"""Computes the unit vector of a vector."""
function normalize(v::Vec2)
    m = mag(v)
    x = v.x
    y = v.y
    Vec2(x / m, y / m)
end

function normalize(v::Vec3)
    m = mag(v)
    x = v.x
    y = v.y
    z = v.z
    Vec3(x / m, y / m, z / m)
end

# TODO: Will add rounding method after I better understand the rounding mechanism in Julia.
# Rounding is more complex than I originally thought.

end # module Vectors

